smolyay/surrogate.py

```diff
@@ -209,14 +219,16 @@ def gradient(self, x):
                                     self.grid.basis_functions):
                 if self.dimension > 1:
                     term = numpy.prod(
-                        [f.derivative(x) if dim == ni else f(x) for dim, (x, f)
-                         in enumerate(zip(x_scaled, basis))])
+                            [basis[dim].derivative(x_scaled[...,dim]) 
+                             if dim == ni else basis[dim](x_scaled[...,dim]) 
+                             for dim in range(len(basis))], axis=0)
                 else:
-                    term = basis.derivative(x_scaled[0])
+                    term = basis.derivative(x_scaled)
+                
                 value += coeff*term
-            gradient.append(value)
-
-        return gradient[0] if self.dimension == 1 else gradient
+            gradient.append(numpy.squeeze(value))
+        gradient = numpy.array(gradient,copy=False).transpose()
+        return numpy.squeeze(gradient)
```

Comment:
I don't follow what all is going on here, could you give an example of why you need this order of squeeze and transpose? Maybe some of this could be fixed using the suggestion to preallocate above.

Reply:
I was replicating the way is was originally written and so the value is squeezed to avoid a nested array being added to gradient. 

The transpose is added because gradient is supposed to be a matrix with a size num_points x dimension. value will be an array the size of num_points. However, appending adds value to gradient's row when it should be the column, hence the transpose.

The last squeeze is to again collapse the number of dimensions, which is particularly important in the 1D case. Originally the return line was 

`return gradient[0] if self.dimension == 1 else gradient`

which had the effect of preventing gradient from being a nested list for 1D cases. However, this obviously would prevent any gradients being returned for any points except the first one, so it was replaced with squeeze which did the same thing.

The transpose was causing issues if the points were given in multi-dimensional grids (so gradient's shape had more than 2 dimensions), and I pushed a commit fixes this issue.

Reply:
I allocated gradient to a numpy array with removes the need for reshaping completely, but both instances of squeezing are required to pass all the pytests.